core/src/commons/Data.cpp

```diff
@@ -174,22 +175,41 @@ void Data::get_all_values(std::vector<double>& all_values,
    // fill with [0, 1,..., samples.size() - 1]
   std::iota(index.begin(), index.end(), 0);
   // sort index based on the split values (argsort)
-  std::sort(index.begin(), index.end(), [&](const size_t& lhs, const size_t& rhs)
-    {return all_values[lhs] < all_values[rhs];});
+  if (contains_nan()) {
+    // the NaN comparison places all NaNs at the beginning
+    // this will be slightly slower if there are no NaNs, thus the branch
+    std::stable_sort(index.begin(), index.end(), [&](const size_t& lhs, const size_t& rhs) {
+      if (std::isnan(all_values[lhs]) && !std::isnan(all_values[rhs])) {
+        return true;
+      } else {
+        return all_values[lhs] < all_values[rhs];
+      }
+    });
+  } else {
+    std::sort(index.begin(), index.end(), [&](const size_t& lhs, const size_t& rhs)
+      {return all_values[lhs] < all_values[rhs];});
+  }
 
   for (size_t i = 0; i < samples.size(); i++) {
     sorted_samples[i] = samples[index[i]];
     all_values[i] = get(sorted_samples[i], var);
   }
 
-  all_values.erase(unique(all_values.begin(), all_values.end()), all_values.end());
+  all_values.erase(unique(all_values.begin(), all_values.end(), [&](const double& lhs, const double& rhs) {
+    if (std::isnan(lhs) && std::isnan(rhs)) {
```

Comment:
I think this could be written as
```
return lhs == rhs || (std::isnan(lhs) && std::isnan(rhs))
```

The same suggestion applies to the `stable_sort` call above.